src/auth.zig

```diff
@@ -5,23 +5,137 @@ const std = @import("std");
 const Allocator = std.mem.Allocator;
 
 pub const Auth = @This();
+pub const User = @import("auth/user.zig");
 
-user: User = User{},
+provider: AnyAuth,
+current_user: ?User = null,
 
-pub const User = struct {
-    username: []const u8 = "invalid username",
+pub const Error = error{
+    UnknownUser,
+    Unauthenticated,
+    NotProvided,
 };
 
 pub fn valid(a: Auth) bool {
-    _ = a;
-    return true;
+    return a.provider.valid() catch false;
 }
 
-pub fn validOrError(a: Auth) !void {
-    if (!a.valid()) return error.Unauthenticated;
+pub fn requireValid(a: Auth) !void {
+    if (a.current_user == null or !a.valid()) return error.Unauthenticated;
 }
 
-pub fn currentUser(a: Auth, alloc: Allocator) !User {
-    _ = alloc;
-    return a.user;
+pub fn Provider(T: type) type {
+    return struct {
+        const Self = @This();
+        ctx: T,
+
+        pub fn init(ctx: T) Self {
+            return .{
+                .ctx = ctx,
+            };
+        }
+
+        /// TODO document the implications of non consttime function
+        pub fn lookupUser(self: *const Self, user_id: []const u8) Error!User {
+            return try self.ctx.lookupUser(user_id);
+        }
+
+        pub fn any(self: *const Self) AnyAuth {
+            return .{
+                .ctx = self,
+                .lookup_user = lookupUserUntyped,
+            };
+        }
+
+        fn lookupUserUntyped(self: *const anyopaque, user_id: []const u8) Error!User {
+            const typed: *const T = @ptrCast(self);
+            return typed.lookupUser(user_id);
+        }
+    };
+}
+
+/// Type Erased Version of an auth provider
+pub const AnyAuth = struct {
+    ctx: *const anyopaque,
+    lookup_user: ?LookupUserFn = null,
+    valid_: ?ValidFn = null,
```

Comment:
This is never set any where, maybe this is intentional for now?

Reply:
It has to exist for everything to compile with the existing code in srctree (which is my primary test bed), I can plumb it up, or ignore it? The former requires me to add a srctree implementation first. 

I can't just deleted it because srctree needs to compile, and I don't wanna delete the code to just re add it.